### ISSUE TITLE
Crash on OSX 10.12

### ISSUE THREAD

**Issue by raph68** [cavaliercoder/dmidecode-osx#3]
dmidecode
# dmidecode 3.0
Getting SMBIOS data from Apple SMBIOS service.
SMBIOS 2.4 present.
61 structures occupying 2888 bytes.

Trace/BPT trap: 5
No way to get any further, any ideas?

**Comment by silkimen** [comment]
+1

**Comment by ghost** [comment]
+1, using a hackintosh

**Comment by cavaliercoder** [comment]
Thanks for posting. I don't have access to 10.12 at the moment. Will work on it when I get some bandwidth, but PRs are very welcome.

**Comment by ghost** [comment]
I'll try to debug the issue and post more details

**Comment by ghost** [comment]
OK, found the bug. The data is read correctly, but the program crashes when freeing a CFMutableDictionaryRef. I suppose it's because it's a reference and maybe it's pointless to free a reference because the system does it..?

dmidecode.c:4581
Comment that line.
![debug](https://cloud.githubusercontent.com/assets/3678359/22367711/58ffe9fe-e483-11e6-835a-b39a0aacb268.jpg)

**Comment by cavaliercoder** [comment]
According to the docs for [IORegistryEntryCreateCFProperties](https://developer.apple.com/reference/iokit/1514310-ioregistryentrycreatecfpropertie#parameters), I am meant to free the ref. Though the ownership behaviours of these APIs are a little poorly described. I also note, that the CFRelease call works fine prior to 10.12.

I've added 87d2dc497be8ab5849e50aa4577e24d760324f3f which checks that `IORegistryEntryCreateCFProperties` returns OK. I don't image any issues would have slipped through the next check in the call to `CFDictionaryGetValueIfPresent`, but maybe this will highlight something.

**Comment by cavaliercoder** [comment]
I've also added NULL checks for each CFRelease call in 3d85facb9b706dae200c35793e53c4803dd930b8. I can't see why this would/wouldn't work. It's just precautionary. I'm working pretty hard to get a copy of 10.12 now (corporate SOE restrictions....).

**Comment by cavaliercoder** [comment]
Finally got my hands on Sierra. I can replicate the following:
```
# dmidecode 3.0
Getting SMBIOS data from Apple SMBIOS service.
SMBIOS 2.5 present.
10 structures occupying 453 bytes.

Segmentation fault: 11
```

However, if I run with lldb attached, it runs fine... weird.

**Comment by tmaone** [comment]
I got this working. 

1. The ```CFRelease``` in [#L4587](https://github.com/cavaliercoder/dmidecode-osx/blob/3d85facb9b706dae200c35793e53c4803dd930b8/dmidecode.c#L4587) had to go as it was faulting.

2. As I suspected from the funny behaviour in ```lldb``` vs non-```lldb``` execution, it boils down to the signed binary. ```lldb``` is a signed binary. The freshly compiled ```dmidecode``` is not. I originally thought that it would need entitlements too, but this is not the case. Signing that binary with any available developer profile fixes the issue and then program returns appropriate output.
